lewis_emulators/aldn1000/interfaces/stream_interface.py

```diff
@@ -0,0 +1,166 @@
+    def format_data(self, float):
+        # Restrict a float to only the first 5 characters
+        return '{formatted_value:.5s}'.format(formatted_value=str(float))
+
+    @if_connected
+    def set_pump(self, address, action):
+        self.device.address = address
+        if action == 'STP':
+            self.device.pump = 'STP'
+            if self.device.status == 'P':  # Currently paused
+                self.device.status = 'S'  # Stop
+            elif self.device.status == 'S':
+                pass
+            else:
+                self.device.status = 'P'  # Pause
+        elif action == 'RUN':
+            self.device.pump = 'RUN'
+            if self.device.direction == 'Infusing':
+                self.device.status = 'I'
+            else:
+                self.device.status = 'W'
+        else:
+            print('An error occurred while trying to start/stop the pump')
+        return self.basic_set_response(address)
+
+    @if_connected
+    def get_status(self, address):
+        return self.basic_get_response(address, self.device.status)
+
+    @if_connected
+    def set_pump_on(self, address):
+        self.device.address = address
+        self.device.pump = 'ON'
+
+    @if_connected
+    def set_pump_off(self, address):
+        self.device.address = address
+        self.device.pump = 'OFF'
+
+    @if_connected
+    def get_diameter(self, address):
+        return self.basic_get_response(address, self.format_data(self.device.diameter))
+
+    @if_connected
+    def set_diameter(self, address, diameter):
+        self.device.address = address
+        self.device.diameter = diameter
+        if diameter > 14.00:
+            self.device.volume_units = 'mL'
+        else:
+            self.device.volume_units = 'uL'
+        return self.basic_set_response(address)
+
+    @if_connected
+    def get_volume(self, address):
+        if self.device.volume_units == 'mL':
+            units = 'ML'
+        else:
+            units = 'UL'
+        return self.basic_get_response(address, self.format_data(self.device.volume), units)
+
+    @if_connected
+    def set_volume(self, address, volume):
+        self.device.address = address
+        self.device.volume = volume
+        return self.basic_set_response(address)
+
+    @if_connected
+    def get_volume_dispensed(self, address):
+        data = 'I{infused}W{withdrawn}{units}'.format(infused=self.format_data(self.device.volume_infused),
+                                                      withdrawn=self.format_data(self.device.volume_withdrawn),
+                                                      units=self.device.volume_units)
+        return self.basic_get_response(address, data)
+
+    @if_connected
+    def get_direction(self, address):
+        return self.basic_get_response(address, self.device.direction)
+
+    @if_connected
+    def set_direction(self, address, direction):
+        self.device.address = address
+        if direction == 'REV':  # Reverse
```

Comment:
This logic can also be covered by the state machine

Reply:
Created state machine.